Statsd protocol compatibility

Koilanetroc opened the pull request:
In this PR I added compatibility with [statsd protocol](https://github.com/b/statsd_spec) with tags to be able to send metrics directly from statsd clients to vmagent or directly to VM.
For example its compatible with [statsd-instrument](https://github.com/Shopify/statsd-instrument) and [dogstatsd-ruby](https://github.com/DataDog/dogstatsd-ruby) gems

Related issues: #5052, #206, #4600

hagen1778 commented:
Thanks for PR! Will be waiting for it to be promoted from draft before giving review

Koilanetroc commented:
@hagen1778, I finished it and now its ready for review, please have a look

Koilanetroc commented:
I see the failing test, but not sure if it relates to my changes, because it also fails on master locally for me. Rebased on actual master

Koilanetroc commented:
Rebased on current master, had conflicts in `docs/keyConcepts.md`

guybortnikov commented:
Hey, I was wondering if there have been any updates on this feature? 
I am eagerly waiting for this feature.

Koilanetroc commented:
@guybortnikov, I'm waiting for maintainers review, please check for last updates in this [thread](https://victoriametrics.slack.com/archives/CGZF1H6L9/p1705337394383279).  

They previously said they were interested in this functionality, but probably didn't have time to proceed with this one

f41gh7 commented:
Hello, thanks for the PR! 

 I think, it misses the most critical part of statsd server - aggregation. 


Most probably, we can adapt current streaming aggregation for statsd protocol. In this case, it must be avaliable only for vmagent and requires pre-configuration of streaming aggregation.

 Basic logic:
1) add new `PushFunc` into `app/vmagent/remotewrite` package. It must add timeseries only to aggregators, ignoring direct push to the remote write
2) add new aggregators if needed for statsd types ( like `counter_sample` aggregator, that performs sampling for counters).
3) adds documentation and examples for it. 


 cc @valyala @hagen1778

Koilanetroc commented:
@f41gh7, rebased on last master and fixed incompatibility with it, now it works well.  

> I think, it misses the most critical part of statsd server - aggregation.

Could you explain what you mean? Streaming aggregation works well when vmagent accepts statsd data(tested locally with enabled aggregation). As I understand streaming aggregation works regardless of the way data was ingested. Or we are talking about different aggregations?

Koilanetroc commented:
Regarding statsd sample rate - its accepted, but ignored. Not sure if there is a need to implement sampling on vmagent, but let me know if you think differently

Koilanetroc commented:
Hi @f41gh7, @valyala, @hagen1778
Just rebased on current master, resolved a few docs conflicts. Please let me know if you are sill interested in vmagent statsd protocol or maybe your plans have changed already. I see number of 👍  grows, so it seems to be a wanted feature

f41gh7 commented:
> Hi @f41gh7, @valyala, @hagen1778 Just rebased on current master, resolved a few docs conflicts. Please let me know if you are sill interested in vmagent statsd protocol or maybe your plans have changed already. I see number of 👍 grows, so it seems to be a wanted feature

 Hello, sry for delay, we're going to merge it soon with a few additions:
1) respond with an error, if streaming aggregation is disabled.
2) add a new flag `statsd.disableEnforcedAggregation=false`, which disables a previous check.

 It must protect database from overload by direct client pushes with high metrics resolution.

f41gh7 reviewed (approved):
LGTM

f41gh7 commented:
@Koilanetroc May I ask you to squad and sign commits? 

 Please check the following guide https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

Koilanetroc commented:
@f41gh7, sure, squashed and signed

f41gh7 commented:
Thanks for contribution!

valyala commented:
@Koilanetroc , please see the follow-up commit f4051dd1e0abf3c327cb75fe026736feeee0f735 , which is created by running `make docs-sync` for syncing `README.md` with the corresponding files in `docs` folder.

valyala commented:
FYI, this pull request has been reverted in 2da7dfc7548cf8dda523d012e39eb7059289f933 . Please read commit description for the reason of the revert.

Koilanetroc commented:
@valyala, regarding your reasoning described [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/206#issuecomment-2207394456).  

Did I get it right that the main point is that this implementation is not a really statsd drop-in replacement since a developer needs also to create a streaming aggregation config in order to call it a full replacement?

If so, shouldn't we just remove `drop-in` from docs and mention that it is rather only a statsd format compatibility? And later add ability to enable via flag a streaming aggregation config which copies statsd behaviour.

I mean it sounds a bit radical to revert this functionality only because people may misunderstand it and assume that there is already implemented some aggregation by default. 

---

Also do you see critical obstacles in adding a default aggregation config copying statsd behaviour? I'm currently not sure at what moment streaming aggregation kicks in, but can't this be solved by adding an additional flag like `copyStatsdAggregationConfig` and passing to streaming aggregation some predefined config which copies statsd behaviour?
